app/tournament/tournament.js

```diff
@@ -0,0 +1,101 @@
+import * as player from "./player";
+import * as match from "./match";
+import * as utils from "../common/utils";
+
+export class Tournament {
+    constructor() {
+        this.players = [];
+        this.matches = [];
+    }
+
+    addPlayer(player) {
+        if (this.existsAlready(player)) {
+            throw new TournamentError(`Duplicate player: "${player.name}".`);
+        }
+        this.players.push(player);
+    }
+
+    existsAlready(player) {
+        for (var p of this.players) {
+            if (p.name.toLowerCase() === player.name.toLowerCase()) {
+                return true;
+            }
+        };
+        return false;
+    }
+
+    generate() {
+        this.checkNumberOfPlayers();
+        this.generateEmptyMatches();
+        this.generateTournamentBracket();
+        this.pairUpPlayers();
+    }
+
+    checkNumberOfPlayers() {
+        if (!this.enaughPlayers()) {
+            throw new TournamentError(`Too few players: ${this.players.length}`);
+        }
+        if (!this.hasProperNumberOfPlayers()){
+            this.fillUpWithGhosts();
+        }
+    }
+
+    enaughPlayers() {
+        return this.players.length >= 2;
+    }
+
+    hasProperNumberOfPlayers() {
+        return Number.isInteger(Math.log2(this.players.length));
+    }
+
+    fillUpWithGhosts() {
+        var goodNumber = parseInt(Math.pow(10, this.players.length.toString(2).length), 2);
+        var i = 1;
+        while (this.players.length < goodNumber) {
+            this.addPlayer(new player.Player(`Ghost${i++}`, 1));
+        }
+    }
+
+    generateEmptyMatches() {
+        for (var i=0; i<this.totalMatches(); i++) {
+            this.matches.push(new match.Match());
+        }
+    }
```

Comment:
The idea after ghost players is that matches where they are involved are also ghost matches, so the other player must promote to the next round and the ghost match removed.
So draft cannot be randomly, because 2 ghost players cannot be paired together. Using ranks and giving ghost players the lowest will do the trick.